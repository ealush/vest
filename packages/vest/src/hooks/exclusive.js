import hasOwnProperty from 'hasOwnProperty';

import asArray from 'asArray';
import context from 'ctx';
import { ERROR_HOOK_CALLED_OUTSIDE } from 'hookErrors';
import isStringValue from 'isStringValue';
import {
  EXCLUSION_ITEM_TYPE_TESTS,
  EXCLUSION_ITEM_TYPE_GROUPS,
} from 'runnableTypes';
import throwError from 'throwError';

/**
 * @type {String} Exclusion group name: only.
 */
const EXCLUSION_GROUP_NAME_ONLY = 'only';

/**
 * @type {String} Exclusion group name: skip.
 */
const EXCLUSION_GROUP_NAME_SKIP = 'skip';

/**
 * Adds fields to a specified exclusion group.
 * @param {String} exclusionGroup   To add the fields to.
 * @param {String} itemType         Whether the item is a group or a test.
 * @param {String[]|String} item    A field name or a list of field names.
 */
const addTo = (exclusionGroup, itemType, item) => {
  const ctx = context.use();
  if (!item) {
    return;
  }

  if (!ctx) {
    throwError(`${exclusionGroup} ${ERROR_HOOK_CALLED_OUTSIDE}`);
    return;
  }

  asArray(item).forEach(itemName => {
    if (!isStringValue(itemName)) {
      return null;
    }

    ctx.exclusion[itemType][itemName] =
      exclusionGroup === EXCLUSION_GROUP_NAME_ONLY;
  });
};

/**
 * Adds a field or multiple fields to inclusion group.
 * @param {String[]|String} item Item to be added to inclusion group.
 */
const only = item =>
  addTo(EXCLUSION_GROUP_NAME_ONLY, EXCLUSION_ITEM_TYPE_TESTS, item);

only.group = item =>
  addTo(EXCLUSION_GROUP_NAME_ONLY, EXCLUSION_ITEM_TYPE_GROUPS, item);

/**
 * Adds a field or multiple fields to exclusion group.
 * @param {String[]|String} item Item to be added to exclusion group.
 */
const skip = item =>
  addTo(EXCLUSION_GROUP_NAME_SKIP, EXCLUSION_ITEM_TYPE_TESTS, item);

skip.group = item =>
  addTo(EXCLUSION_GROUP_NAME_SKIP, EXCLUSION_ITEM_TYPE_GROUPS, item);

/**
 * Checks whether a certain test profile excluded by any of the exclusion groups.
 * @param {String} fieldName    Field name to test.
 * @param {VestTest}            Test Object reference.
 * @returns {Boolean}
 */
const isExcluded = testObject => {
  const { fieldName, groupName } = testObject;

  const { exclusion } = context.use();

  const keyTests = exclusion[EXCLUSION_ITEM_TYPE_TESTS];
  const testValue = keyTests[fieldName];

  // if test is skipped
  // no need to proceed
  if (testValue === false) {
    return true;
  }

  const isTestIncluded = testValue === true;

  // If inside a group
  if (groupName) {
    if (isGroupExcluded(groupName)) {
      return true; // field excluded by group

      // if group is `only`ed
    } else if (exclusion[EXCLUSION_ITEM_TYPE_GROUPS][groupName] === true) {
      if (isTestIncluded) {
        return false;
      }

      // If there is _ANY_ `only`ed test (and we already know this one isn't)
      if (hasIncludedTests(keyTests)) {
        return true; // Excluded implicitly
      }

      return keyTests[fieldName] === false;
    }
  }

  // if field is only'ed
  if (isTestIncluded) {
    return false;
  }

  // If there is _ANY_ `only`ed test (and we already know this one isn't) return true
  // Otherwise return false
  return hasIncludedTests(keyTests);
};

/**
 * Checks if context has included tests
 * @param {Object} keyTests Object containing included and excluded tests
 * @returns {boolean}
 */
const hasIncludedTests = keyTests => {
  for (const test in keyTests) {
    if (keyTests[test] === true) {
      return true; // excluded implicitly
    }
  }
  return false;
};

/**
 * Checks whether a given group is excluded from running.
 * @param {String} groupName
 * @return {Boolean}
 */
const isGroupExcluded = groupName => {
  const { exclusion } = context.use();
  const keyGroups = exclusion[EXCLUSION_ITEM_TYPE_GROUPS];

  const groupPresent = hasOwnProperty(keyGroups, groupName);

  // When group is either only'ed or skipped
  if (groupPresent) {
    // Return true if group is skipped and false if only'ed
    return keyGroups[groupName] === false;
  }

  // Group is not present
  for (const group in keyGroups) {
    // If any other group is only'ed
    if (keyGroups[group] === true) {
      return true;
    }
  }

  return false;
};

export { only, skip, isExcluded, isGroupExcluded };
