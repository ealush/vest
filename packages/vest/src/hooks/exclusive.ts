import asArray from 'asArray';
import hasOwnProperty from 'hasOwnProperty';
import isStringValue from 'isStringValue';
import throwError from 'throwError';

import VestTest from 'VestTest';
import ctx from 'ctx';
import { ERROR_HOOK_CALLED_OUTSIDE } from 'hookErrors';

type TExclusionItemType = string | string[] | undefined;

/**
 * Adds a field or multiple fields to inclusion group.
 */
export function only(item: TExclusionItemType): void {
  return addTo('only', 'tests', item);
}

only.group = (item: TExclusionItemType) => addTo('only', 'groups', item);

/**
 * Adds a field or multiple fields to exclusion group.
 * @param {String[]|String} item Item to be added to exclusion group.
 */
export function skip(item: TExclusionItemType): void {
  return addTo('skip', 'tests', item);
}

skip.group = (item: TExclusionItemType) => addTo('skip', 'groups', item);

/**
 * Checks whether a certain test profile excluded by any of the exclusion groups.
 * @param {VestTest}            Test Object reference.
 * @returns {Boolean}
 */
export function isExcluded(testObject: VestTest): boolean {
  const { fieldName, groupName } = testObject;

  const context = ctx.use();
  const exclusion = context!.exclusion;

  const keyTests = exclusion!.tests;
  const testValue = keyTests[fieldName];

  // if test is skipped
  // no need to proceed
  if (testValue === false) {
    return true;
  }

  const isTestIncluded = testValue === true;

  // If inside a group
  if (groupName) {
    if (isGroupExcluded(groupName)) {
      return true; // field excluded by group

      // if group is `only`ed
    } else if (exclusion!.groups[groupName] === true) {
      if (isTestIncluded) {
        return false;
      }

      // If there is _ANY_ `only`ed test (and we already know this one isn't)
      if (hasIncludedTests(keyTests)) {
        return true; // Excluded implicitly
      }

      return keyTests[fieldName] === false;
    }
  }

  // if field is only'ed
  if (isTestIncluded) {
    return false;
  }

  // If there is _ANY_ `only`ed test (and we already know this one isn't) return true
  // Otherwise return false
  return hasIncludedTests(keyTests);
}

/**
 * Checks whether a given group is excluded from running.
 * @param {String} groupName
 * @return {Boolean}
 */
export function isGroupExcluded(groupName: string): boolean {
  const context = ctx.use();
  const exclusion = context!.exclusion;
  const keyGroups = exclusion!.groups;

  const groupPresent = hasOwnProperty(keyGroups, groupName);

  // When group is either only'ed or skipped
  if (groupPresent) {
    // Return true if group is skipped and false if only'ed
    return keyGroups[groupName] === false;
  }

  // Group is not present
  for (const group in keyGroups) {
    // If any other group is only'ed
    if (keyGroups[group] === true) {
      return true;
    }
  }

  return false;
}

/**
 * Adds fields to a specified exclusion group.
 */
function addTo(
  exclusionGroup: 'only' | 'skip',
  itemType: 'tests' | 'groups',
  item: TExclusionItemType
) {
  const context = ctx.use();
  if (!item) {
    return;
  }

  if (!context) {
    throwError(`${exclusionGroup} ${ERROR_HOOK_CALLED_OUTSIDE}`);
    return;
  }

  asArray(item).forEach((itemName: string): void => {
    if (!isStringValue(itemName)) {
      return;
    }

    context!.exclusion![itemType][itemName] = exclusionGroup === 'only';
  });
}

/**
 * Checks if context has included tests
 */
function hasIncludedTests(keyTests: Record<string, boolean>): boolean {
  for (const test in keyTests) {
    if (keyTests[test] === true) {
      return true; // excluded implicitly
    }
  }
  return false;
}
