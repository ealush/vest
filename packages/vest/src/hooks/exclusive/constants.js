/**
 * @type {String} Exclusion group name: only.
 */
export const EXCLUSION_GROUP_NAME_ONLY = 'only';

/**
 * @type {String} Exclusion group name: skip.
 */
export const EXCLUSION_GROUP_NAME_SKIP = 'skip';

export const EXCLUSION_ITEM_TYPE_TESTS = 'tests';
export const EXCLUSION_ITEM_TYPE_GROUPS = 'groups';
