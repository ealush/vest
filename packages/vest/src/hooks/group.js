import context from 'ctx';
import isFunction from 'isFunction';
import isStringValue from 'isStringValue';
import throwError from 'throwError';

/**
 * Runs a group callback.
 * @param {string} groupName
 * @param {Function} tests
 */
const group = (groupName, tests) => {
  if (!isStringValue(groupName)) {
    throwError(
      `group initialization error. Expected "${groupName}" to be a string.`
    );
  }

  if (!isFunction(tests)) {
    throwError(
      `group initialization error. Expected "${tests}" to be a function.`
    );
  }

  // Running with the context applied
  context.bind({ groupName }, tests)();
};

export default group;
