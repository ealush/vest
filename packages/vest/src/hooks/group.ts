import isFunction from 'isFunction';
import isStringValue from 'isStringValue';
import throwError from 'throwError';

import context from 'ctx';

/**
 * Runs a group callback.
 */
export default function group(groupName: string, tests: () => any): void {
  if (!isStringValue(groupName)) {
    throwGroupError(groupName);
  }

  if (!isFunction(tests)) {
    throwGroupError(tests);
  }

  // Running with the context applied
  context.bind({ groupName }, tests)();
}

function throwGroupError(value: unknown) {
  throwError(
    __DEV__
      ? `group initialization error. Expected "${value}" to be a string.`
      : 'group name must be a string'
  );
}
