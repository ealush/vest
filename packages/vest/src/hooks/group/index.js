import context from '../../core/context';
import validateSuiteParams from '../../lib/validateSuiteParams';
import { isGroupExcluded } from '../exclusive';

/**
 * Registers a group in state.
 * @param {string} groupName
 */
const registerGroup = groupName => {
  const { stateRef } = context.use();
  stateRef.patch(state => {
    const nextState = { ...state };
    nextState.groups[groupName] = state.groups[groupName] || {};
    return nextState;
  });
};

/**
 * Runs a group callback.
 * @param {string} groupName
 * @param {Function} tests
 */
const group = (groupName, tests) => {
  // TODO: Replace with something local
  validateSuiteParams('group', groupName, tests);

  const { stateRef } = context.use();
  const state = stateRef.current();

  if (!isGroupExcluded(state, groupName)) {
    registerGroup(groupName);
  }

  context.run(
    {
      groupName,
    },
    () => tests()
  );
};

export default group;
