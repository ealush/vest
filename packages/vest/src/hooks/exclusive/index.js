import getSuiteState from '../../core/state/getSuiteState';
import patch from '../../core/state/patch';
import singleton from '../../lib/singleton';
import throwError from '../../lib/throwError';
import { ERROR_HOOK_CALLED_OUTSIDE } from '../constants';
import { GROUP_NAME_ONLY, GROUP_NAME_SKIP } from './constants';

/**
 * Adds fields to a specified group.
 * @param {String} group            To add the fields to.
 * @param {String[]|String} item    A field name or a list of field names.
 */
const addTo = (group, item) => {
  const ctx = singleton.useContext();

  if (!item) {
    return;
  }
  if (ctx?.suiteId === undefined) {
    throwError(`${group} ${ERROR_HOOK_CALLED_OUTSIDE}`);
    return;
  }

  patch(ctx.suiteId, state => {
    const nextState = { ...state };
    [].concat(item).forEach(fieldName => {
      if (typeof fieldName === 'string') {
        nextState.exclusive[group] = nextState.exclusive[group] || {};
        nextState.exclusive[group][fieldName] = true;
      }
    });
    return nextState;
  });
};

/**
 * Adds a field or multiple fields to inclusion group.
 * @param {String[]|String} item Item to be added to inclusion group.
 */
export const only = item => addTo(GROUP_NAME_ONLY, item);

/**
 * Adds a field or multiple fields to exclusion group.
 * @param {String[]|String} item Item to be added to exclusion group.
 */
export const skip = item => addTo(GROUP_NAME_SKIP, item);

/**
 * Checks whether a certain field name is excluded by any of the exclusion groups.
 * @param {String} fieldName    FieldN name to test.
 * @returns {Boolean}
 */
export const isExcluded = fieldName => {
  const ctx = singleton.useContext();

  if (ctx?.suiteId === undefined) {
    return false;
  }

  const state = getSuiteState(ctx.suiteId);

  const group = ctx.groupName;

  // If we're inside a group but what we're checking
  // is a different field.
  if (group && fieldName !== group) {
    // If the group is `only`ed
    if (state.exclusive?.[GROUP_NAME_ONLY]?.[group]) {
      return !!state.exclusive?.[GROUP_NAME_SKIP]?.[fieldName]; // excluded if current field is skipped
    }
  }

  // If field is skipped
  if (state.exclusive?.[GROUP_NAME_SKIP]?.[fieldName]) {
    return true; // excluded
  }

  // If there is _ANY_ `only`ed group
  if (state.exclusive?.[GROUP_NAME_ONLY]) {
    // If the current field is `only`ed
    if (state.exclusive[GROUP_NAME_ONLY]?.[fieldName]) {
      return false; // Not excluded
    }

    // There's an `only`ed field, but it's not this one
    return true; // excluded
  }

  return false; // Not excluded
};
