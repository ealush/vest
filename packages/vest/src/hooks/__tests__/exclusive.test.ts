import faker from 'faker';

import mock from '../../../testUtils/mock';
import testDummy from '../../../testUtils/testDummy';

import VestTest from 'VestTest';
import context from 'ctx';
import { isExcluded, isGroupExcluded } from 'exclusive';
import group from 'group';
import { ERROR_HOOK_CALLED_OUTSIDE } from 'hookErrors';
import * as vest from 'vest';


let res, res1;

describe('exclusive hooks', () => {
  let test1, test2, test3;

  beforeEach(() => {
    test1 = new VestTest(faker.lorem.word(), jest.fn());
    test2 = new VestTest(faker.lorem.slug(), jest.fn());
    test3 = new VestTest(faker.random.word(), jest.fn());
  });

  test('isExcluded should respect group exclusion', () => {
    let testObject;
    let testObject1;
    const validate = vest.create(() => {
      vest.skip.group('group_1');

      testObject = testDummy(vest).failing();

      group('group_1', () => {
        testObject1 = testDummy(vest).failing();
      });

      res = isExcluded(testObject);
      res1 = isExcluded(testObject1);
    });

    validate();

    expect(res).toBe(false);
    expect(res1).toBe(true);
  });

  describe('`only` hook', () => {
    describe('string input', () => {
      test('isExcluded returns false for included field', () => {
        vest.create(() => {
          vest.only(test1.fieldName);
          res = isExcluded(test1);
        })();
        expect(res).toBe(false);
      });

      test('isGroupExcluded returns false for included groups', () => {
        vest.create(() => {
          vest.only('group_name');

          group('group_name', () => {
            res = isGroupExcluded('group_name');
          });
        })();
        expect(res).toBe(false);
      });

      test('isExcluded returns true for non included field', () => {
        vest.create(() => {
          vest.only(test1.fieldName);
          res1 = isExcluded(test1);
          res = isExcluded(test2);
        })();
        expect(res1).toBe(false);
        expect(res).toBe(true);
      });

      test('isGroupExcluded returns true for non included group', () => {
        vest.create(() => {
          vest.only.group('group_1');

          group('group_1', jest.fn());
          group('group_2', jest.fn());

          res1 = isGroupExcluded('group_1');
          res = isGroupExcluded('group_2');
        })();

        expect(res1).toBe(false);
        expect(res).toBe(true);
      });
    });

    describe('array input', () => {
      test('isExcluded returns false for included field', () => {
        vest.create(() => {
          vest.only([test1.fieldName, test2.fieldName]);
          res = isExcluded(test1);
          res1 = isExcluded(test2);
        })();
        expect(res).toBe(false);
        expect(res1).toBe(false);
      });

      test('isGroupExcluded returns false for included groups', () => {
        vest.create(() => {
          vest.only.group(['group_1', 'group_2']);

          group('group_1', jest.fn());
          group('group_2', jest.fn());
          group('group_3', jest.fn());

          res = [
            isGroupExcluded('group_1'),
            isGroupExcluded('group_2'),
            isGroupExcluded('group_3'),
          ];
        })();
        expect(res).toEqual([false, false, true]);
      });

      test('isExcluded returns true for non included field', () => {
        vest.create(() => {
          res = [isExcluded(test1), isExcluded(test2), isExcluded(test3)];
          vest.only([test1.fieldName, test2.fieldName]);
          res1 = [isExcluded(test1), isExcluded(test2), isExcluded(test3)];
        })();
        expect(res).toEqual([false, false, false]);
        expect(res1).toEqual([false, false, true]);
      });

      test('isGroupExcluded returns true for non included groups', () => {
        vest.create(() => {
          vest.only.group(['group_1', 'group_2']);

          group('group_3', jest.fn());
          res = [
            isGroupExcluded('group_1'),
            isGroupExcluded('group_2'),
            isGroupExcluded('group_3'),
          ];
        })();
        expect(res).toEqual([false, false, true]);
      });
    });
  });

  describe('`skip` hook', () => {
    describe('string input', () => {
      test('isExcluded returns true for excluded field', () => {
        vest.create(() => {
          vest.skip(test1.fieldName);
          res = isExcluded(test1);
        })();
        expect(res).toBe(true);
      });

      test('isGroupExcluded returns true for excluded groups', () => {
        vest.create(() => {
          vest.skip.group('group_1');
          res = isGroupExcluded('group_1');
          res1 = isGroupExcluded('group_2');
        })();

        expect(res).toBe(true);
        expect(res1).toBe(false);
      });

      test('isExcluded returns false for non excluded field', () => {
        vest.create(() => {
          vest.skip(test1.fieldName);
          res = isExcluded(vest.test(test2, jest.fn()));
        })();
        expect(res).toBe(false);
      });
    });

    test('isGroupExcluded returns false for tests in non excluded groups', () => {
      vest.create(() => {
        vest.skip('group_1');
        res = isExcluded(vest.test('field_1', jest.fn()));
      })();
      expect(res).toBe(false);
    });

    describe('array input', () => {
      test('isExcluded returns true for excluded field', () => {
        vest.create(() => {
          vest.skip([test1.fieldName, test2.fieldName]);

          res = isExcluded(test1);
          res1 = isExcluded(test2);
        })();
        expect(res).toBe(true);
        expect(res1).toBe(true);
      });

      test('isGroupExcluded returns true for excluded groups', () => {
        vest.create(() => {
          vest.skip.group(['group_1', 'group_2']);
          res = [isGroupExcluded('group_1'), isGroupExcluded('group_2')];
        })();
        expect(res).toEqual([true, true]);
      });

      test('isExcluded returns false for non included field', () => {
        vest.create(() => {
          vest.skip([test1.fieldName, test2.fieldName]);
          res = isExcluded(test3);
        })();
        expect(res).toBe(false);
      });

      test('isGroupExcluded returns false for non excluded groups', () => {
        vest.create(() => {
          vest.skip(['group_1', 'group_2']);
          res = isGroupExcluded('group_3');
        })();
        expect(res).toEqual(false);
      });
    });
  });

  describe('Error handling', () => {
    let mockThrowError, hooks;

    beforeEach(() => {
      mockThrowError = mock('throwError');
      hooks = require('exclusive');
    });

    describe.each([['only', 'skip']])('%s', hook => {
      describe('When called outside of a suite', () => {
        it('Should throw an error', () => {
          hooks[hook](faker.random.word());
          expect(mockThrowError.mock.calls[0][0]).toContain(
            ERROR_HOOK_CALLED_OUTSIDE
          );
        });
      });
    });
  });
});

describe('isExcluded', () => {
  let exclusion;

  const runIsExcluded = (exclusion, testObject: VestTest) =>
    context.run({}, ctx => {
      Object.assign(ctx.exclusion, exclusion);
      const res = isExcluded(testObject);

      return res;
    });

  const genTest = (fieldName: string, groupName?: string) =>
    new VestTest(fieldName, jest.fn(), {
      groupName,
    });
  describe('skip', () => {
    beforeEach(() => {
      exclusion = { tests: { field_1: false, field_2: false } };
    });
    it('returns true for skipped field', () => {
      expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(true);
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
        true
      );
    });
    it('returns false for non skipped field', () => {
      expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_4', 'group_1'))).toBe(
        false
      );
    });
  });
  describe('only', () => {
    beforeEach(() => {
      exclusion = { tests: { field_1: true, field_2: true } };
    });
    it('returns false for included field', () => {
      expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
        false
      );
    });
    it('returns true for non included field', () => {
      expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
      expect(runIsExcluded(exclusion, genTest('field_4', 'group_1'))).toBe(
        true
      );
    });
  });
  describe('only+skip', () => {
    beforeEach(() => {
      exclusion = {
        tests: { field_1: true, field_2: true, field_3: false, field_4: false },
      };
    });
    it('returns false for included tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
        false
      );
    });
    it('returns true excluded tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
      expect(runIsExcluded(exclusion, genTest('field_4', 'group_1'))).toBe(
        true
      );
    });
    it('returns true for non included field', () => {
      expect(runIsExcluded(exclusion, genTest('field_5'))).toBe(true);
      expect(runIsExcluded(exclusion, genTest('field_6', 'group_1'))).toBe(
        true
      );
    });
  });
  describe('skip.group', () => {
    beforeEach(() => {
      exclusion = {
        groups: {
          group_1: false,
          group_2: false,
        },
      };
    });

    it('Returns true for tests in skipped group', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
        true
      );
    });
    it('Returns false for tests in non skipped groups', () => {
      expect(runIsExcluded(exclusion, genTest('field_3', 'group_3'))).toBe(
        false
      );
      expect(runIsExcluded(exclusion, genTest('field_4', 'group_4'))).toBe(
        false
      );
    });
    it('Returns false for tests outside of any group', () => {
      expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_4'))).toBe(false);
    });
  });
  describe('only.group', () => {
    beforeEach(() => {
      exclusion = {
        groups: {
          group_1: true,
          group_2: true,
        },
      };
    });

    it('returns false for tests in included groups', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
        false
      );
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_2'))).toBe(
        false
      );
    });

    it('returns true for groups in non included groups', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
        true
      );
    });

    it('returns false for tests outside of any group', () => {
      expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(false);
    });
  });

  describe('only.group + only', () => {
    beforeEach(() => {
      exclusion = {
        groups: {
          group_1: true,
          group_2: true,
        },
        tests: { field_1: true, field_2: true },
      };
    });

    it('returns false for included tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(false);
    });

    it('returns false for included tests in included groups', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
        false
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
        false
      );
    });

    it('returns true for included test in non included group', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
        true
      );
    });

    it('returns true for non included test in included group', () => {
      expect(runIsExcluded(exclusion, genTest('field_3', 'group_1'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_4', 'group_2'))).toBe(
        true
      );
    });

    it('returns true for non included tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
      expect(runIsExcluded(exclusion, genTest('field_4'))).toBe(true);
    });
  });

  describe('skip.group + only', () => {
    beforeEach(() => {
      exclusion = {
        groups: {
          group_1: false,
          group_2: false,
        },
        tests: { field_1: true, field_2: true },
      };
    });

    it('returns true for tests in excluded groups', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_1'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_3', 'group_1'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_2'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_3', 'group_2'))).toBe(
        true
      );
    });

    it('returns false for tests in included tests in non excluded groups', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
        false
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_3'))).toBe(
        false
      );
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_4'))).toBe(
        false
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
        false
      );
    });

    it('returns true for non included tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(true);
      expect(runIsExcluded(exclusion, genTest('field_4'))).toBe(true);
    });

    it('returns false for included tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(false);
    });
  });

  describe('only.group + skip', () => {
    beforeEach(() => {
      exclusion = {
        groups: {
          group_1: true,
          group_2: true,
        },
        tests: { field_1: false, field_2: false },
      };
    });

    it('returns true for excluded tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_1'))).toBe(true);
      expect(runIsExcluded(exclusion, genTest('field_2'))).toBe(true);
    });
    it('returns false for non excluded tests', () => {
      expect(runIsExcluded(exclusion, genTest('field_3'))).toBe(false);
      expect(runIsExcluded(exclusion, genTest('field_4'))).toBe(false);
    });
    it('returns true for excluded test in included group', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_1'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_2'))).toBe(
        true
      );
    });
    it('returns true for excluded test in non included group', () => {
      expect(runIsExcluded(exclusion, genTest('field_1', 'group_3'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_2', 'group_4'))).toBe(
        true
      );
    });
    it('returns false for non excluded test in included group', () => {
      expect(runIsExcluded(exclusion, genTest('field_3', 'group_1'))).toBe(
        false
      );
      expect(runIsExcluded(exclusion, genTest('field_4', 'group_2'))).toBe(
        false
      );
    });
    it('returns true for non excluded test in non included group', () => {
      expect(runIsExcluded(exclusion, genTest('field_3', 'group_3'))).toBe(
        true
      );
      expect(runIsExcluded(exclusion, genTest('field_4', 'group_4'))).toBe(
        true
      );
    });
  });
});
