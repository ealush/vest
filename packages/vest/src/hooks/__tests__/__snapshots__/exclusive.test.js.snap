// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exclusive hooks \`skipWhen\` hook When \`shouldSkip\` is \`false\` Should produce correct validation result 1`] = `
Object {
  "done": [Function],
  "errorCount": 3,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {
    "group": Object {
      "field2": Object {
        "errorCount": 1,
        "testCount": 1,
        "warnCount": 0,
      },
    },
  },
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": undefined,
  "testCount": 3,
  "tests": Object {
    "field1": Object {
      "errorCount": 1,
      "testCount": 1,
      "warnCount": 0,
    },
    "field2": Object {
      "errorCount": 1,
      "testCount": 1,
      "warnCount": 0,
    },
    "field3": Object {
      "errorCount": 1,
      "testCount": 1,
      "warnCount": 0,
    },
  },
  "warnCount": 0,
}
`;
